modules/mapper-extras/src/main/java/org/elasticsearch/index/query/SourceIntervalsSource.java

```diff
@@ -0,0 +1,196 @@
+
+            @Override
+            public int nextDoc() throws IOException {
+                return doNext(approximation.nextDoc());
+            }
+
+            @Override
+            public int advance(int target) throws IOException {
+                return doNext(approximation.advance(target));
+            }
+
+            private int doNext(int doc) throws IOException {
+                while (doc != NO_MORE_DOCS && setIterator(doc) == false) {
+                    doc = approximation.nextDoc();
+                }
+                return doc;
+            }
+
+            private boolean setIterator(int doc) {
+                try {
+                    final List<Object> values = valueFetcher.apply(doc);
+                    final LeafReaderContext singleDocContext = createSingleDocLeafReaderContext(field, values);
+                    in = SourceIntervalsSource.this.in.intervals(field, singleDocContext);
+                    final boolean isSet = in != null && in.nextDoc() != NO_MORE_DOCS;
+                    assert isSet == false || in.docID() == 0;
+                    return isSet;
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
+
+            @Override
+            public int start() {
+                return in.start();
+            }
+
+            @Override
+            public int end() {
+                return in.end();
+            }
+
+            @Override
+            public int gaps() {
+                return in.gaps();
+            }
+
+            @Override
+            public int nextInterval() throws IOException {
+                return in.nextInterval();
+            }
+
+            @Override
+            public float matchCost() {
+                // a high number since we need to parse the _source
+                return 10_000;
+            }
+
+        };
+    }
+
+    @Override
+    public IntervalMatchesIterator matches(String field, LeafReaderContext ctx, int doc) throws IOException {
+        final CheckedIntFunction<List<Object>, IOException> valueFetcher = valueFetcherProvider.apply(ctx);
+        final List<Object> values = valueFetcher.apply(doc);
+        final LeafReaderContext singleDocContext = createSingleDocLeafReaderContext(field, values);
+        return in.matches(field, singleDocContext, 0);
+    }
+
+    @Override
+    public void visit(String field, QueryVisitor visitor) {
+        in.visit(field, visitor);
+    }
+
+    @Override
+    public int minExtent() {
+        return in.minExtent();
+    }
+
+    @Override
+    public Collection<IntervalsSource> pullUpDisjunctions() {
+        return Collections.singleton(this);
+    }
+
+    @Override
+    public int hashCode() {
+        // Not using matchesProvider and valueFetcherProvider, which don't identify this source but are only used to avoid scanning linearly
+        // through all documents
+        return Objects.hash(in, indexAnalyzer);
```

Comment:
The index analyzer should be immutable for an open index, I think? So I'm not sure that it needs to be included here or in equals.

Reply:
Agreed. I'm including it because I want to avoid making too many assumptions about how this class is used by Elasticsearch. Is it fine with you?